hibernate-core/src/main/java/org/hibernate/type/EntityType.java

```diff
@@ -262,6 +263,22 @@ public final Object nullSafeGet(
 		return resolve( hydrate( rs, names, session, owner ), session, owner );
 	}
 
```

Comment:
I don't remember the details for `SpecialOneToOneType` (which extends `OneToOneType`), so I'm not sure if the following changes are safe for `SpecialOneToOneType`.

I believe the reason why OneToOneType#nullSafeSet did nothing is because in the unidirectional (non-mappedBy) case, the parent entity ID itself is the foreign key. The entity ID is already bound to the query elsewhere, so it shouldn't be bound again. We need to make sure that adding this block does not break this case.

I would need to see more tests of these other cases to be sure it doesn't break anything.